src/OrbitGl/App.cpp

```diff
@@ -1881,6 +1882,150 @@ Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModu
   return chained_result_future;
 }
 
+Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModuleItself(
+    const ModuleIdentifier& module_id, uint64_t module_file_size) {
+  ORBIT_SCOPE_FUNCTION;
+  ORBIT_CHECK(std::this_thread::get_id() == main_thread_id_);
+
+  const auto it = symbol_files_currently_being_retrieved_.find(module_id);
+  if (it != symbol_files_currently_being_retrieved_.end()) {
+    return it->second;
+  }
+
+  // Note that the bullet points in the ErrorMessage constructed by this function are indented (by
+  // one level). This is because the caller of this method integrates this ErrorMessage into an
+  // ErrorMessage that also has bullet points (with no indentation, i.e., at top level).
+
+  auto find_in_cache_or_locally =
+      [this, module_id, module_file_size]() -> ErrorMessageOr<CanceledOr<std::filesystem::path>> {
+    std::string error_message;
+    {
+      auto object_in_cache = symbol_helper_.FindObjectInCache(module_id.file_path,
+                                                              module_id.build_id, module_file_size);
+      if (object_in_cache.has_value()) {
+        ORBIT_LOG("Found module file \"%s\" itself in cache", module_id.file_path);
+        return {object_in_cache.value()};
+      }
+      error_message += absl::StrFormat("\n  * Could not find module file itself in cache: %s",
+                                       object_in_cache.error().message());
+    }
+    if (absl::GetFlag(FLAGS_local)) {
+      auto verify_object_file_result = orbit_symbols::VerifyObjectFile(
+          module_id.file_path, module_id.build_id, module_file_size);
+      if (verify_object_file_result.has_value()) {
+        ORBIT_LOG("Found module file \"%s\" itself locally", module_id.file_path);
+        return module_id.file_path;
+      }
+      error_message += "\n  * Could not find module file itself locally.";
+    }
+    return ErrorMessage{error_message};
+  };
+
+  Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> find_in_cache_or_locally_future =
+      thread_pool_->Schedule(find_in_cache_or_locally);
+
+  if (download_disabled_modules_.contains(module_id.file_path)) {
+    return find_in_cache_or_locally_future;
+  }
```

Comment:
Having this here, means the module_id does not get added to the `symbol_files_currently_being_retrieved_` list. This is probably not a big issue, since only local things happen and those are hopefullly pretty fast. But nevertheless, I would feel better if you move this at the beginning of `retrieve_from_instance`